staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -173,11 +176,28 @@ func pingTimeoutSeconds() int {
 	return ret
 }
 
-func configureHTTP2Transport(t *http.Transport) error {
-	t2, err := http2.ConfigureTransports(t)
+func configureHTTP2Transport(tInner *http.Transport) (*http.Transport, error) {
+	// We want to clone the original transport and disable HTTP/2 on the clone.  Per
+	// godoc, if TLSNextProto is not nil, HTTP/2 support is not enabled.  Hence, we
+	// do this dance with save/restore.
+	innerTLSNextProto := tInner.TLSNextProto
+	tInner.TLSNextProto = map[string]func(string, *tls.Conn) http.RoundTripper{}
+	tOuter := tInner.Clone()
+	tInner.TLSNextProto = innerTLSNextProto
+
+	// Now that we have an outer transport same as incoming but without HTTP/2, we
+	// configure HTTP/2 on inner and set it to handle "https" on the outer.
+	t2, err := http2.ConfigureTransports(tInner)
 	if err != nil {
-		return err
+		return nil, err
 	}
+
+	// Because the stdlib transport tries to send upgrades over HTTP/2, we wrap
+	// it in our own h2RoundTripper which returns http.ErrSkipAltProtocol on
+	// upgrade requests. This causes the tOuter transport to handle the request
+	// (which only does HTTP/1.1_
```

Comment:
this is the part that I'm more worried about, suddenly the same client will handle 2 transports, one for upgraded requests and other for the rest, ...  I have not clear the consequences of this
why do we need to automate this? why not fix the spdy rountripper only and let the users configured it independently?